diff --git a/src/models/actor_critic.py b/src/models/actor_critic.py
index c1f24ec..964db2d 100644
--- a/src/models/actor_critic.py
+++ b/src/models/actor_critic.py
@@ -25,22 +25,6 @@ class actor_critic(nn.Module):
 		if(continuous):
 			self.actor = continuous_net(hidden_dim, state_dim, action_dim, num_layers, dropout)
 			self.critic = critic(hidden_dim, state_dim, num_layers, dropout)
-			"""
-			self.critic = nn.Sequential(
-				layer_init(nn.Linear(np.array(state_dim).prod(), 64)),
-				nn.Tanh(),
-				layer_init(nn.Linear(64, 64)),
-				nn.Tanh(),
-				layer_init(nn.Linear(64, 1), std=1.0),
-			)
-			self.actor_alt = nn.Sequential(
-				layer_init(nn.Linear(np.array(state_dim).prod(), 64)),
-				nn.Tanh(),
-				layer_init(nn.Linear(64, 64)),
-				nn.Tanh(),
-				layer_init(nn.Linear(64, np.prod(action_dim)), std=0.01),
-			)
-			"""
 			self.actor_logstd = nn.Parameter(torch.zeros(1, np.prod(action_dim)))
 		else:
 			self.actor = discrete_net(hidden_dim, state_dim, action_dim, num_layers, dropout)
diff --git a/src/run.py b/src/run.py
index 9a1c57c..5187e5c 100644
--- a/src/run.py
+++ b/src/run.py
@@ -26,7 +26,7 @@ def plot_curves(arr_list, legend_list, x_indices, color_list, ylabel, fig_title)
 
 if __name__=='__main__':
 	parser = argparse.ArgumentParser()
-	parser.add_argument('-id', '--gym_id', type=str, help='Id of the environment that we will use', default='CartPole-v1')
+	parser.add_argument('-id', '--gym_id', type=str, help='Id of the environment that we will use', default='HalfCheetah-v4')
 	parser.add_argument('-s', '--seed', type=float, help='Seed for experiment', default=1.0)
 	parser.add_argument('-ns', '--num_steps', type=int, help='Number of steps that the environment should take', default=128)
 	parser.add_argument('-gae', '--gae', type=bool, help='Generalized Advantage Estimation flag', default=True)
@@ -45,7 +45,7 @@ if __name__=='__main__':
 	parser.add_argument('-na', '--norm_adv', type=bool, help='Normalize advantage estimates', default=True)
 	parser.add_argument('-p', '--capture_video', type=bool, help='Whether to capture the video or not', default=False)
 	parser.add_argument('-d', '--hidden_dim', type=int, help='Hidden dimension of the neural networks in the actor critic', default=64)
-	parser.add_argument('-c', '--continuous', type=bool, help='Is the action space continuous',default=False)
+	parser.add_argument('-c', '--continuous', type=bool, help='Is the action space continuous',default=True)
 	parser.add_argument('-lr', '--learning_rate', type=float, help='Learning rate for our agent', default=2.5e-4)
 	parser.add_argument('-exp', '--exp_name', type=str, help='Experiment name', default='CartPole PPO')
 	parser.add_argument('-nl', '--num_layers', type=int, help='The number of layers in our actor and critic', default=2)
diff --git a/src/wandb/latest-run b/src/wandb/latest-run
index b419769..0e755f3 120000
--- a/src/wandb/latest-run
+++ b/src/wandb/latest-run
@@ -1 +1 @@
-run-20231204_132639-kgrwu2by
\ No newline at end of file
+run-20231204_232728-9rmf1qlr
\ No newline at end of file
